examples/BooksellerFancy.hs

```diff
@@ -0,0 +1,90 @@
+{-# LANGUAGE BlockArguments #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE TemplateHaskell #-}
+
+{-
+# Maximally fancy bookseller
+-}
+
+module BooksellerFancy where
+
+import CLI
+import Choreography
+import Choreography.Network.Http
+import Data (deliveryDateOf, priceOf)
+import System.Environment
+
+$(mkLoc "buyer")
+$(mkLoc "seller")
+
+proof1 :: Subset ("buyer" ': supporters) ("buyer" ': "seller" ': supporters)
+proof1 = undefined
+
+proof2 :: Subset '["buyer", "seller"] ("buyer" ': "seller" ': supporters)
+proof2 = undefined
+
+-- | `bookseller` is a choreography that implements the bookseller protocol.
+-- This version takes a choreography `mkDecision` that implements the decision making process.
+bookseller :: (KnownSymbols supporters) =>
+  (Located '["buyer"] Int -> Choreo ("buyer" ': supporters) (CLI m) (Located '["buyer"] Bool)) ->
+  Choreo ("buyer" ': "seller" ': supporters) (CLI m) ()
+bookseller mkDecision = do
+  database <- seller `_locally` getInput "Enter the book database (for `Read`):"
+  title <- (buyer, getstr "Enter the title of the book to buy:") -~> seller @@ nobody
+
+  -- the seller checks the price of the book and sends it to the buyer
+  price <- (seller, \un -> return $ priceOf (un seller database) (un seller title)) ~~> buyer @@ nobody
+
+  -- the buyer and supporters make a decision using the `mkDecision` choreography
+  decision <- enclave proof1 $ mkDecision price
+
+  -- get ride of the extra Located (TODO: make a helper function for this?)
+  decision <- locally buyer $ \un -> return (un buyer (un buyer decision))
+
+  -- if the buyer decides to buy the book, the seller sends the delivery date to the buyer
+  enclave proof2 $ broadcast' First (First, decision) >>= \case
+    True -> do
+      deliveryDate <- (seller, \un -> return $ deliveryDateOf (un seller database) (un seller title)) ~~> buyer @@ nobody
```

Comment:
There's also `*~>`, which would spare you the `return`. I'm afraid there's _not_ a clear logic to which helper functions exist or what they're called though :(